File: app/src/main/java/home/oleg/popularmovies/domain/usecases/DetailInteractor.java
package home.oleg.popularmovies.domain.usecases;

import java.util.List;

import home.oleg.popularmovies.data.entities.ReviewResponse;
import home.oleg.popularmovies.data.entities.VideosResponse;
import home.oleg.popularmovies.domain.MovieRepository;
import home.oleg.popularmovies.domain.UseCaseObserver;
import home.oleg.popularmovies.presentation.model.MovieViewModel;

/**
 * Created by Oleg Sheliakin on 06.08.2017.
 * Contact me by email - olegsheliakin@gmail.com
 */

public class DetailInteractor {

    private final GetReviewsUseCase getReviewsUseCase;
    private final GetTrailersUseCase getTrailersUseCase;
    private final MovieRepository movieRepository;

    public DetailInteractor(GetReviewsUseCase getReviewsUseCase,
                            GetTrailersUseCase getTrailersUseCase,
                            MovieRepository movieRepository) {
        this.getReviewsUseCase = getReviewsUseCase;
        this.getTrailersUseCase = getTrailersUseCase;
        this.movieRepository = movieRepository;
    }

    public void executeFetchReviews(UseCaseObserver<List<ReviewResponse.Result>> observer, Integer id) {
        getReviewsUseCase.execute(observer, id);
    }

    public void executeFetchTrailers(UseCaseObserver<List<VideosResponse.Result>> observer, Integer id) {
        getTrailersUseCase.execute(observer, id);
    }

    public boolean isFavourite(Integer id) {
        return movieRepository.isFavourite(id);
    }

    public void addToFavourite(MovieViewModel movieViewModel) {
        movieRepository.save(movieViewModel);
    }

    public void deleteFromFavourite(Integer id) {
        movieRepository.delete(id);
    }

}

File: app/src/main/java/home/oleg/popularmovies/domain/usecases/DetailInteractor.kt
package home.oleg.popularmovies.domain.usecases

import home.oleg.popularmovies.data.entities.ReviewResponse
import home.oleg.popularmovies.data.entities.VideosResponse
import home.oleg.popularmovies.domain.FavoriteMovieRepository
import home.oleg.popularmovies.domain.MovieRepository
import home.oleg.popularmovies.domain.entities.Movie
import io.reactivex.Completable
import io.reactivex.Observable
import javax.inject.Inject

/**
 * Created by Oleg Sheliakin on 06.08.2017.
 * Contact me by email - olegsheliakin@gmail.com
 */

class DetailInteractor @Inject constructor(
        private val getReviewsUseCase: GetReviewsUseCase,
        private val getTrailersUseCase: GetTrailersUseCase,
        private val favoriteMovieRepository: FavoriteMovieRepository) {

    fun fetchReviews(id: Long): Observable<List<ReviewResponse.Result>> {
        return getReviewsUseCase.execute(id)
    }

    fun fetchTrailers(id: Long): Observable<List<VideosResponse.Result>> {
        return getTrailersUseCase.execute(id)
    }

    fun addToFavourite(id: Long): Completable {
        return favoriteMovieRepository.addFavorite(id)
    }

    fun deleteFromFavourite(id: Long): Completable {
        return favoriteMovieRepository.removeFavorite(id)
    }

}
